presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4

```diff
@@ -540,6 +541,20 @@ STRING
     : '\'' ( ~'\'' | '\'\'' )* '\''
     ;
 
+// note that the following is really not precisely binaryLiteral, but
+// "something that looks like binaryLiteral". here we also include
+// the form X'any string', which is a valid typeConstructor syntax.
+// but we also disallow users from declaring a type 'X', so the expression
+// X'any string' is thrown out later in parsing.
+// the rationale for this is that when user writes X'some string', he's most
+// likely trying to write a binary literal, so processing the error inside
+// the binaryLiteral parsing route would give more informative error msg
+// in this syntax rule, we also allow odd-number of digits, which is not exactly
+// the sql spec, but we check for that later in the parsing flow
+BINARY_LITERAL
+    :  'X\'' (~'\'' | '\'\'')* '\''
```

Comment:
this current setup again comes to our compromise between describing the
true grammar rules fully in the .g4 file vs doing part of the grammar check
in the .g4 and checking for full syntax in the BinaryLiteral() constructor.
 the current assumption is that if someone wants to write a binaryLiteral
but accidentally put a "'" inside the string, this thing would parse as a
binaryLiteral and we could give out more concrete error message in that
route, vs directing the parsing to the type_constructor/genericLiteral
route and just say "ok u gave a X 'something' , u must be meaning
binaryLiteral". --- of course we could also add that ~=m/'/ check in the
GenericLiteral() constructor check, but then the same logic starts to
diverge more and more into 2 places


On Sat, Dec 19, 2015 at 11:54 AM, Martin Traverso <notifications@github.com>
wrote:

> In presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4
> <https://github.com/facebook/presto/pull/3997#discussion_r48096133>:
>
> > @@ -540,6 +541,20 @@ STRING
> >      : '\'' ( ~'\'' | '\'\'' )* '\''
> >      ;
> >
> > +// note that the following is really not precisely binaryLiteral, but
> > +// "something that looks like binaryLiteral". here we also include
> > +// the form X'any string', which is a valid typeConstructor syntax.
> > +// but we also disallow users from declaring a type 'X', so the expression
> > +// X'any string' is thrown out later in parsing.
> > +// the rationale for this is that when user writes X'some string', he's most
> > +// likely trying to write a binary literal, so processing the error inside
> > +// the binaryLiteral parsing route would give more informative error msg
> > +// in this syntax rule, we also allow odd-number of digits, which is not exactly
> > +// the sql spec, but we check for that later in the parsing flow
> > +BINARY_LITERAL
> > +    :  'X\'' (~'\'' | '\'\'')* '\''
>
> This is incorrect. Binary literals cannot contain escaped quotes. It
> should be just:
>
>  'X\'' (~'\'')* '\''
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/presto/pull/3997/files#r48096133>.
>